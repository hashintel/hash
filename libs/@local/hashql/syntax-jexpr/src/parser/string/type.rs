use hashql_ast::node::{
    id::NodeId,
    r#type::{
        IntersectionType, StructField, StructType, TupleField, TupleType, Type, TypeKind, UnionType,
    },
};
use winnow::{
    ModalResult, Parser as _,
    combinator::{
        alt, delimited, dispatch, opt, peek, repeat, separated, separated_foldl1, separated_pair,
        terminated,
    },
    error::ParserError,
    token::any,
};

use super::{
    combinator::{separated_boxed1, ws},
    context::Input,
    ident::parse_ident,
};
use crate::parser::string::path::parse_path;

fn parse_type_infer<'heap, 'span, 'source, E>(
    input: &mut Input<'heap, 'span, 'source>,
) -> ModalResult<Type<'heap>, E>
where
    E: ParserError<Input<'heap, 'span, 'source>>,
{
    let context = input.state;

    "_".with_span()
        .map(|(_, span)| Type {
            id: NodeId::PLACEHOLDER,
            span: context.span(span),
            kind: TypeKind::Infer,
        })
        .parse_next(input)
}

fn parse_type_tuple_field<'heap, 'span, 'source, E>(
    input: &mut Input<'heap, 'span, 'source>,
) -> ModalResult<TupleField<'heap>, E>
where
    E: ParserError<Input<'heap, 'span, 'source>>,
{
    let context = input.state;

    parse_type
        .with_span()
        .map(|(r#type, span)| TupleField {
            id: NodeId::PLACEHOLDER,
            span: context.span(span),
            r#type,
        })
        .parse_next(input)
}

fn parse_type_tuple<'heap, 'span, 'source, E>(
    input: &mut Input<'heap, 'span, 'source>,
) -> ModalResult<Type<'heap>, E>
where
    E: ParserError<Input<'heap, 'span, 'source>>,
{
    let context = input.state;

    let empty = "()".with_span().map(|(_, span)| TupleType {
        id: NodeId::PLACEHOLDER,
        span: context.span(span),
        fields: context.heap.empty_slice(),
    });

    let fields = delimited(
        ws("("),
        (
            repeat(1.., terminated(parse_type_tuple_field, ws(","))),
            opt(parse_type_tuple_field),
        )
            .map(|(types, last): (Vec<_>, _)| {
                let required_length = types.len() + usize::from(last.is_some());

                let mut vec = context.heap.vec(Some(required_length));

                vec.extend(types);
                vec.extend(last);

                vec.into_boxed_slice()
            }),
        ws(")"),
    )
    .with_span()
    .map(|(fields, span)| TupleType {
        id: NodeId::PLACEHOLDER,
        span: context.span(span),
        fields,
    });

    alt((empty, fields))
        .map(|tuple| Type {
            id: NodeId::PLACEHOLDER,
            span: tuple.span,
            kind: TypeKind::Tuple(tuple),
        })
        .parse_next(input)
}

fn parse_type_struct_field<'heap, 'span, 'source, E>(
    input: &mut Input<'heap, 'span, 'source>,
) -> ModalResult<StructField<'heap>, E>
where
    E: ParserError<Input<'heap, 'span, 'source>>,
{
    let context = input.state;

    separated_pair(parse_ident, ws(":"), parse_type)
        .with_span()
        .map(|((name, r#type), span)| StructField {
            id: NodeId::PLACEHOLDER,
            span: context.span(span),
            name,
            r#type,
        })
        .parse_next(input)
}

fn parse_type_struct<'heap, 'span, 'source, E>(
    input: &mut Input<'heap, 'span, 'source>,
) -> ModalResult<Type<'heap>, E>
where
    E: ParserError<Input<'heap, 'span, 'source>>,
{
    let context = input.state;

    let empty = "(:)".with_span().map(|(_, span)| StructType {
        id: NodeId::PLACEHOLDER,
        span: context.span(span),
        fields: context.heap.empty_slice(),
    });

    let fields = delimited(
        ws("("),
        terminated(
            separated_boxed1(context.heap, parse_type_struct_field, ws(",")),
            opt(ws(",")),
        ),
        ws(")"),
    )
    .with_span()
    .map(|(fields, span)| StructType {
        id: NodeId::PLACEHOLDER,
        span: context.span(span),
        fields,
    });

    alt((empty, fields))
        .map(|r#struct| Type {
            id: NodeId::PLACEHOLDER,
            span: r#struct.span,
            kind: TypeKind::Struct(r#struct),
        })
        .parse_next(input)
}

fn parse_type_paren<'heap, 'span, 'source, E>(
    input: &mut Input<'heap, 'span, 'source>,
) -> ModalResult<Type<'heap>, E>
where
    E: ParserError<Input<'heap, 'span, 'source>>,
{
    delimited(ws("("), parse_type, ws(")")).parse_next(input)
}

fn parse_type_atom<'heap, 'span, 'source, E>(
    input: &mut Input<'heap, 'span, 'source>,
) -> ModalResult<Type<'heap>, E>
where
    E: ParserError<Input<'heap, 'span, 'source>>,
{
    let mut path = parse_path.map(|path| Type {
        id: NodeId::PLACEHOLDER,
        span: path.span,
        kind: TypeKind::Path(path),
    });

    dispatch! {peek(any);
        '_' => parse_type_infer,
        '(' => alt((parse_type_paren, parse_type_tuple, parse_type_struct)),
        _ => path
    }
    .parse_next(input)
}

fn parse_type_union<'heap, 'span, 'source, E>(
    input: &mut Input<'heap, 'span, 'source>,
) -> ModalResult<Type<'heap>, E>
where
    E: ParserError<Input<'heap, 'span, 'source>>,
{
    let context = input.state;

    separated(1.., parse_type_atom, ws("|"))
        .with_span()
        .map(|(mut parsed, span): (Vec<_>, _)| {
            if parsed.len() == 1 {
                return parsed.pop().unwrap_or_else(|| unreachable!());
            }

            let span = context.span(span);

            let mut types = context.heap.vec(Some(parsed.len()));
            types.extend(parsed);

            Type {
                id: NodeId::PLACEHOLDER,
                span,
                kind: TypeKind::Union(UnionType {
                    id: NodeId::PLACEHOLDER,
                    span,
                    types: types.into_boxed_slice(),
                }),
            }
        })
        .parse_next(input)
}

fn parse_type_intersection<'heap, 'span, 'source, E>(
    input: &mut Input<'heap, 'span, 'source>,
) -> ModalResult<Type<'heap>, E>
where
    E: ParserError<Input<'heap, 'span, 'source>>,
{
    let context = input.state;

    separated(1.., parse_type_union, ws("&"))
        .with_span()
        .map(|(mut parsed, span): (Vec<_>, _)| {
            if parsed.len() == 1 {
                return parsed.pop().unwrap_or_else(|| unreachable!());
            }

            let span = context.span(span);

            let mut types = context.heap.vec(Some(parsed.len()));
            types.extend(parsed);

            Type {
                id: NodeId::PLACEHOLDER,
                span,
                kind: TypeKind::Intersection(IntersectionType {
                    id: NodeId::PLACEHOLDER,
                    span,
                    types: types.into_boxed_slice(),
                }),
            }
        })
        .parse_next(input)
}

/// ```abnf
/// infer = "_"
/// tuple = "()" / "(" +(type ",") ?(type) ")"
/// struct = "(:)" / "(" ident ":" type *("," ident ":" type) ?"," ")"
/// paren = "(" type ")"
/// atom = path / tuple / struct / infer / paren
/// union = atom *("|" atom)
/// intersection = union *("&" union)
/// type = intersection
/// ```
pub(crate) fn parse_type<'heap, 'span, 'source, E>(
    input: &mut Input<'heap, 'span, 'source>,
) -> ModalResult<Type<'heap>, E>
where
    E: ParserError<Input<'heap, 'span, 'source>>,
{
    ws(parse_type_intersection).parse_next(input)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::parser::string::test::{bind_parser, test_cases};

    bind_parser!(SyntaxDump; fn parse_type_test(parse_type));

    // Tests for individual parser components
    test_cases!(parse_type_test;
        // Atom types
        infer_type("_") => "Infer type",
        simple_path("Int") => "Simple path type",
        qualified_path("Std::String") => "Qualified path",
        nested_qualified_path("Std::Collections::HashMap") => "Nested qualified path",

        // Tuple types
        empty_tuple("()") => "Empty tuple",
        single_element_tuple("(Int,)") => "Single-element tuple with trailing comma",
        multi_element_tuple("(Int, String, Boolean)") => "Multi-element tuple",

        // Struct types
        empty_struct("(:)") => "Empty struct",
        single_field_struct("(name: String)") => "Single-field struct",
        multi_field_struct("(name: String, age: Int)") => "Multi-field struct",

        // Union types
        simple_union("Int | String") => "Simple union type",
        multi_union("Int | String | Boolean") => "Multiple union type",

        // Intersection types
        simple_intersection("Int & String") => "Simple intersection type",
        multi_intersection("Int & String & Boolean") => "Multiple intersection type",
    );

    // Tests for composite structures and nesting
    test_cases!(parse_type_test;
        // Nested structures
        nested_tuple("((Int, String), Boolean)") => "Nested tuple type",
        nested_struct("(person: (name: String, age: Int))") => "Nested struct type",
        struct_with_tuple_field("(data: (Int, String))") => "Struct with tuple field",
        tuple_with_struct_field("((name: String), Int)") => "Tuple with struct field",

        // Whitespace handling
        tuple_with_whitespace("( Int , String )") => "Tuple with whitespace",
        struct_with_whitespace("( name : String , age : Int )") => "Struct with whitespace",
        union_with_whitespace(" Int | String ") => "Union with whitespace",
        intersection_with_whitespace(" Int & String ") => "Intersection with whitespace",
    );

    // Tests for operator precedence and complex expressions
    test_cases!(parse_type_test;
        union_of_atoms("Int | String | Boolean") => "Union of atom types",
        intersection_of_atoms("Int & String & Boolean") => "Intersection of atom types",
        union_of_intersections("Int & String | Boolean & Char") => "Union of intersections",
        intersection_of_unions("Int | String & Boolean | Char") => "Intersection of unions",
        mixed_precedence("Int & (String | Boolean) & Char") => "Mixed precedence with parentheses",
        deeply_nested("(((Int | String) & Boolean) | Char)") => "Deeply nested expression",
    );

    // Error cases
    test_cases!(parse_type_test;
        unclosed_tuple("(Int, String") => "Unclosed tuple",
        unclosed_struct("(name: String, age: Int") => "Unclosed struct",
        missing_field_type("(name: )") => "Missing field type",
        invalid_field_separator("(name String)") => "Invalid field separator",
        unclosed_parenthesized("(Int & String") => "Unclosed parenthesized type",
        incomplete_union("Int |") => "Incomplete union",
        incomplete_intersection("Int &") => "Incomplete intersection",
        invalid_operator("Int $ String") => "Invalid operator",
        double_operator("Int || String") => "Double union operator",
        missing_type("") => "Missing type",
    );

    // Tests for common real-world type patterns
    test_cases!(parse_type_test;
        generic_path("Container<T>") => "Generic type",
        generic_with_multiple_params("Map<Key, Value>") => "Generic with multiple parameters",
        nested_generics("Container<Wrapper<T>>") => "Nested generic types",
        optional_type("Option<Value>") => "Optional type pattern",
        complex_entity("(id: ID, attrs: (name: String, metadata: (created: Timestamp, modified: Timestamp | Null)))") => "Complex entity type with nested fields",
    );
}
