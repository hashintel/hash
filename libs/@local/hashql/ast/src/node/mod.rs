//! Core Abstract Syntax Tree node types for HashQL.
//!
//! This module defines the fundamental building blocks of the HashQL Abstract
//! Syntax Tree (AST). The AST represents the structure of a HashQL program after
//! parsing, capturing all the language constructs in a hierarchical tree structure.
//!
//! ## Module Organization
//!
//! The node module is organized into several submodules:
//!
//! - [`expr`]: Expression nodes representing computations that produce values
//! - [`generic`]: Generic type parameter and argument definitions
//! - [`id`]: Node identifier types and utilities
//! - [`path`]: Path expressions for variable references and qualified names
//! - [`type`]: Type system representations including struct, tuple, and other types
//!
//! ## Memory Management
//!
//! All AST nodes use arena allocation through the `'heap` lifetime parameter,
//! which references the memory region where nodes are allocated. This approach
//! improves performance by minimizing allocations and simplifies memory management
//! during parsing and analysis.
//!
//! ## Node Identification
//!
//! Each node in the AST has a unique [`id::NodeId`] that can be used to track
//! the node through various processing stages. Nodes also include a span identifier that points to
//! the source location of the node.
//!
//! ## AST Structure
//!
//! The AST is structured as a tree with expressions ([`expr::Expr`]) at the core.
//! Other node types like types ([`type::Type`]) and paths ([`path::Path`]) are
//! used as components within expressions or on their own for declarations.
//!
//! The entire tree represents a complete HashQL program or query, ready for
//! analysis, transformation, and evaluation.
pub mod expr;
pub mod generic;
pub mod id;
pub mod path;
pub mod r#type;
