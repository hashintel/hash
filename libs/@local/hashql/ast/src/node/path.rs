use hashql_core::{span::SpanId, symbol::Ident};

use super::{generic::GenericArgument, id::NodeId};
use crate::heap;

/// A segment in a path expression.
///
/// Represents a single component of a qualified path, consisting of an identifier
/// and optional generic type arguments. Path segments are connected to form
/// complete paths like `module::submodule::item<T>`.
///
/// # Examples
///
/// In a path like `std::collections::HashMap<K, V>`:
/// - `std` is a segment with no generic arguments
/// - `collections` is a segment with no generic arguments
/// - `HashMap` is a segment with generic arguments `K` and `V`
#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub struct PathSegment<'heap> {
    pub id: NodeId,
    pub span: SpanId,

    pub name: Ident,
    /// Type parameters attached to this path segment
    pub arguments: heap::Vec<'heap, GenericArgument<'heap>>,
}

/// A path expression in the HashQL Abstract Syntax Tree.
///
/// Represents a qualified reference to a variable, function, type, or module.
/// Paths can be simple identifiers (like `x`) or qualified names with multiple
/// segments separated by double colons (like `std::collections::HashMap`).
///
/// Path expressions are used for variable references, module access, and type names.
/// They can include generic type arguments applied to specific segments.
///
/// # Examples
///
/// Simple paths:
/// ```text
/// x                     // A local variable
/// length                // A function or variable
/// ```
///
/// Qualified paths:
/// ```text
/// std::collections::HashMap  // An item in a nested module
/// math::sin                  // A function in a module
/// ```
///
/// Paths with generic arguments:
/// ```text
/// Vec<Int>                       // A type with one type argument
/// HashMap<String, User>          // A type with two type arguments
/// graph::query<User, Connection> // A generic function
/// ```
///
/// J-Expr explicit:
/// ```json
/// {"#path", ["std", "collections", "HashMap"]}
/// {"#path", [{"name": "HashMap", "arguments": ["K, V"]}]}
/// ```
#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub struct Path<'heap> {
    pub id: NodeId,
    pub span: SpanId,

    /// Whether the path is rooted (starts with a double colon `::`).
    pub rooted: bool,
    pub segments: heap::Vec<'heap, PathSegment<'heap>>,
}
