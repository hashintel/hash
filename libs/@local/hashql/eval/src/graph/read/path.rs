use alloc::borrow::Cow;

use hash_graph_store::{
    entity::EntityQueryPath,
    filter::{JsonPath, PathToken, QueryPath},
    subgraph::edges::{EdgeDirection, KnowledgeGraphEdgeKind},
};
use hashql_core::{heap::Heap, literal::LiteralKind, symbol::Symbol, value::Value};

pub(crate) trait CompleteQueryPath<'heap>: QueryPath {
    type PartialQueryPath: PartialQueryPath<'heap, QueryPath = Self>;
}

pub(crate) trait PartialQueryPath<'heap>: Sized {
    type QueryPath;

    fn from_field(heap: &'heap Heap, field: Symbol<'heap>) -> Option<Self>;
    fn access_field(self, heap: &'heap Heap, field: Symbol<'heap>) -> Result<Self, Self>;
    fn from_index(heap: &'heap Heap, index: Cow<'_, Value<'heap>>) -> Option<Self>;
    fn access_index(self, heap: &'heap Heap, index: Cow<'_, Value<'heap>>) -> Result<Self, Self>;
    fn finish(self) -> Option<Self::QueryPath>;
}

impl<'heap> CompleteQueryPath<'heap> for EntityQueryPath<'heap> {
    type PartialQueryPath = PartialEntityQueryPath<'heap>;
}

pub(crate) fn traverse_into_field<'heap, P>(
    path: Option<P>,
    heap: &'heap Heap,
    field: Symbol<'heap>,
) -> Result<P, Option<P>>
where
    P: PartialQueryPath<'heap>,
{
    #[expect(clippy::option_if_let_else, reason = "readability")]
    match path {
        Some(path) => path.access_field(heap, field).map_err(Some),
        None => P::from_field(heap, field).ok_or(None),
    }
}

pub(crate) fn traverse_into_index<'heap, P>(
    path: Option<P>,
    heap: &'heap Heap,
    index: Cow<'_, Value<'heap>>,
) -> Result<P, Option<P>>
where
    P: PartialQueryPath<'heap>,
{
    match path {
        Some(path) => path.access_index(heap, index).map_err(Some),
        None => P::from_index(heap, index).ok_or(None),
    }
}

#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
pub(crate) enum PartialEntityIdQueryPath {
    WebId,
    EntityUuid,
    DraftId,
}

impl<'heap> PartialQueryPath<'heap> for PartialEntityIdQueryPath {
    type QueryPath = EntityQueryPath<'heap>;

    fn from_field(heap: &'heap Heap, field: Symbol<'heap>) -> Option<Self> {
        // see: https://linear.app/hash/issue/H-4512/hashql-use-symbol-sym-constants-instead-of-string-literals
        let web_id = heap.intern_symbol("web_id");
        let entity_uuid = heap.intern_symbol("entity_uuid");
        let draft_id = heap.intern_symbol("draft_id");

        if field == web_id {
            Some(Self::WebId)
        } else if field == entity_uuid {
            Some(Self::EntityUuid)
        } else if field == draft_id {
            Some(Self::DraftId)
        } else {
            None
        }
    }

    fn access_field(self, _: &'heap Heap, _: Symbol<'heap>) -> Result<Self, Self> {
        Err(self)
    }

    fn from_index(_: &'heap Heap, _: Cow<'_, Value<'heap>>) -> Option<Self> {
        None
    }

    fn access_index(self, _: &'heap Heap, _: Cow<'_, Value<'heap>>) -> Result<Self, Self> {
        Err(self)
    }

    fn finish(self) -> Option<Self::QueryPath> {
        match self {
            Self::WebId => Some(EntityQueryPath::WebId),
            Self::EntityUuid => Some(EntityQueryPath::Uuid),
            Self::DraftId => Some(EntityQueryPath::DraftId),
        }
    }
}

#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
pub(crate) enum PartialEntityRecordIdPath {
    EntityId(Option<PartialEntityIdQueryPath>),
    EntityEditionId,
}

impl<'heap> PartialQueryPath<'heap> for PartialEntityRecordIdPath {
    type QueryPath = EntityQueryPath<'heap>;

    fn from_field(heap: &'heap Heap, field: Symbol<'heap>) -> Option<Self> {
        // see: https://linear.app/hash/issue/H-4512/hashql-use-symbol-sym-constants-instead-of-string-literals
        let entity_id = heap.intern_symbol("entity_id");
        let entity_edition_id = heap.intern_symbol("entity_edition_id");

        if field == entity_id {
            Some(Self::EntityId(None))
        } else if field == entity_edition_id {
            Some(Self::EntityEditionId)
        } else {
            None
        }
    }

    fn access_field(self, heap: &'heap Heap, field: Symbol<'heap>) -> Result<Self, Self> {
        match self {
            Self::EntityId(path) => traverse_into_field(path, heap, field)
                .map(Some)
                .map(Self::EntityId)
                .map_err(Self::EntityId),
            Self::EntityEditionId => Err(self),
        }
    }

    fn from_index(_: &'heap Heap, _: Cow<'_, Value<'heap>>) -> Option<Self> {
        None
    }

    fn access_index(self, heap: &'heap Heap, index: Cow<'_, Value<'heap>>) -> Result<Self, Self> {
        match self {
            Self::EntityId(path) => traverse_into_index(path, heap, index)
                .map(Some)
                .map(Self::EntityId)
                .map_err(Self::EntityId),
            Self::EntityEditionId => Err(self),
        }
    }

    fn finish(self) -> Option<Self::QueryPath> {
        match self {
            Self::EntityId(Some(partial)) => partial.finish(),
            Self::EntityId(None) => None,
            Self::EntityEditionId => Some(EntityQueryPath::EditionId),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub(crate) enum PartialLinkDataPath {
    LeftEntityId(Option<PartialEntityIdQueryPath>),
    RightEntityId(Option<PartialEntityIdQueryPath>),
}

impl<'heap> PartialQueryPath<'heap> for PartialLinkDataPath {
    type QueryPath = EntityQueryPath<'heap>;

    fn from_field(heap: &'heap Heap, field: Symbol<'heap>) -> Option<Self> {
        let left_entity_id = heap.intern_symbol("left_entity_id");
        let right_entity_id = heap.intern_symbol("right_entity_id");

        if field == left_entity_id {
            Some(Self::LeftEntityId(None))
        } else if field == right_entity_id {
            Some(Self::RightEntityId(None))
        } else {
            None
        }
    }

    fn access_field(self, heap: &'heap Heap, field: Symbol<'heap>) -> Result<Self, Self> {
        match self {
            Self::LeftEntityId(path) => traverse_into_field(path, heap, field)
                .map(Some)
                .map(Self::LeftEntityId)
                .map_err(Self::LeftEntityId),
            Self::RightEntityId(path) => traverse_into_field(path, heap, field)
                .map(Some)
                .map(Self::RightEntityId)
                .map_err(Self::RightEntityId),
        }
    }

    fn from_index(_: &'heap Heap, _: Cow<'_, Value<'heap>>) -> Option<Self> {
        None
    }

    fn access_index(self, heap: &'heap Heap, index: Cow<'_, Value<'heap>>) -> Result<Self, Self> {
        match self {
            Self::LeftEntityId(path) => traverse_into_index(path, heap, index)
                .map(Some)
                .map(Self::LeftEntityId)
                .map_err(Self::LeftEntityId),
            Self::RightEntityId(path) => traverse_into_index(path, heap, index)
                .map(Some)
                .map(Self::RightEntityId)
                .map_err(Self::RightEntityId),
        }
    }

    #[expect(clippy::match_same_arms, reason = "readability")]
    fn finish(self) -> Option<Self::QueryPath> {
        match self {
            Self::LeftEntityId(Some(partial)) => Some(EntityQueryPath::EntityEdge {
                edge_kind: KnowledgeGraphEdgeKind::HasLeftEntity,
                path: Box::new(partial.finish()?),
                direction: EdgeDirection::Outgoing,
            }),
            Self::LeftEntityId(None) => None,

            Self::RightEntityId(Some(partial)) => Some(EntityQueryPath::EntityEdge {
                edge_kind: KnowledgeGraphEdgeKind::HasRightEntity,
                path: Box::new(partial.finish()?),
                direction: EdgeDirection::Outgoing,
            }),
            Self::RightEntityId(None) => None,
        }
    }
}

fn value_as_usize(value: &Value<'_>) -> Option<usize> {
    match value {
        Value::Primitive(LiteralKind::Integer(integer)) => integer.as_usize(),
        Value::Primitive(LiteralKind::Float(float)) if let Some(integer) = float.as_integer() => {
            integer.as_usize()
        }
        Value::Primitive(_)
        | Value::Struct(_)
        | Value::Tuple(_)
        | Value::List(_)
        | Value::Dict(_)
        | Value::Opaque(_) => None,
    }
}

impl<'heap> PartialQueryPath<'heap> for JsonPath<'heap> {
    type QueryPath = Self;

    fn from_field(_: &'heap Heap, field: Symbol<'heap>) -> Option<Self> {
        Some(JsonPath::from_path_tokens(vec![PathToken::Field(
            Cow::Borrowed(field.unwrap()),
        )]))
    }

    fn access_field(mut self, _: &'heap Heap, field: Symbol<'heap>) -> Result<Self, Self> {
        self.push(PathToken::Field(Cow::Borrowed(field.unwrap())));
        Ok(self)
    }

    fn from_index(_: &'heap Heap, index: Cow<Value<'heap>>) -> Option<Self> {
        let index = value_as_usize(&index)?;

        Some(JsonPath::from_path_tokens(vec![PathToken::Index(index)]))
    }

    fn access_index(mut self, _: &'heap Heap, index: Cow<Value<'heap>>) -> Result<Self, Self> {
        let Some(index) = value_as_usize(&index) else {
            return Err(self);
        };

        self.push(PathToken::Index(index));
        Ok(self)
    }

    fn finish(self) -> Option<Self::QueryPath> {
        Some(self)
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub(crate) enum PartialEntityQueryPath<'heap> {
    Id(Option<PartialEntityRecordIdPath>),
    Properties(Option<JsonPath<'heap>>),
    LinkData(Option<PartialLinkDataPath>),
}

impl<'heap> PartialQueryPath<'heap> for PartialEntityQueryPath<'heap> {
    type QueryPath = EntityQueryPath<'heap>;

    fn from_field(heap: &'heap Heap, field: Symbol<'heap>) -> Option<Self> {
        let id = heap.intern_symbol("id");
        let properties = heap.intern_symbol("properties");
        let link_data = heap.intern_symbol("link_data");

        if field == id {
            Some(PartialEntityQueryPath::Id(None))
        } else if field == properties {
            Some(PartialEntityQueryPath::Properties(None))
        } else if field == link_data {
            Some(PartialEntityQueryPath::LinkData(None))
        } else {
            None
        }
    }

    fn access_field(self, heap: &'heap Heap, field: Symbol<'heap>) -> Result<Self, Self> {
        match self {
            Self::Id(path) => traverse_into_field(path, heap, field)
                .map(Some)
                .map(Self::Id)
                .map_err(Self::Id),
            Self::Properties(path) => traverse_into_field(path, heap, field)
                .map(Some)
                .map(Self::Properties)
                .map_err(Self::Properties),
            Self::LinkData(path) => traverse_into_field(path, heap, field)
                .map(Some)
                .map(Self::LinkData)
                .map_err(Self::LinkData),
        }
    }

    fn from_index(_: &'heap Heap, _: Cow<Value<'heap>>) -> Option<Self> {
        None
    }

    fn access_index(self, heap: &'heap Heap, index: Cow<Value<'heap>>) -> Result<Self, Self> {
        match self {
            Self::Id(path) => traverse_into_index(path, heap, index)
                .map(Some)
                .map(Self::Id)
                .map_err(Self::Id),
            Self::Properties(path) => traverse_into_index(path, heap, index)
                .map(Some)
                .map(Self::Properties)
                .map_err(Self::Properties),
            Self::LinkData(path) => traverse_into_index(path, heap, index)
                .map(Some)
                .map(Self::LinkData)
                .map_err(Self::LinkData),
        }
    }

    #[expect(clippy::match_same_arms, reason = "readability")]
    fn finish(self) -> Option<Self::QueryPath> {
        match self {
            Self::Id(Some(partial)) => partial.finish(),
            Self::Id(None) => None,

            Self::Properties(Some(partial)) => {
                Some(EntityQueryPath::Properties(Some(partial.finish()?)))
            }
            Self::Properties(None) => Some(EntityQueryPath::Properties(None)),

            Self::LinkData(Some(partial)) => partial.finish(),
            Self::LinkData(None) => None,
        }
    }
}
