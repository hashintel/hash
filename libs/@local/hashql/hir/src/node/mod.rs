pub mod access;
pub mod branch;
pub mod call;
pub mod closure;
pub mod data;
pub mod graph;
pub mod input;
pub mod kind;
pub mod r#let;
pub mod operation;
pub mod variable;

use hashql_core::{id, span::SpanId};

use self::kind::NodeKind;

id::newtype!(
    /// A unique identifier for nodes in the HashQL High-Level Intermediate Representation (HIR).
    ///
    /// Each node in the HIR has a unique `NodeId` that can be used to track the node
    /// through various processing stages. These IDs are particularly important for
    /// error reporting, cross-referencing between nodes, and maintaining node identity
    /// during transformations.
    ///
    /// The value space is restricted to 0..=0xFFFF_FF00, reserving the last 256 for niches.
    /// As real pattern types are an experimental feature in Rust, these can currently only be
    /// used by directly modifying and accessing the `NodeId`'s internal value.
    pub struct HirId(u32 is 0..=0xFFFF_FF00)
);

impl HirId {
    /// A placeholder ID used during initial parsing.
    ///
    /// When parsing, we initially give all AST nodes this placeholder ID. After resolving special
    /// forms, the nodes are renumbered to have small indices.
    pub const PLACEHOLDER: Self = <Self as id::Id>::MAX;
}

/// A node in the HashQL High-Level Intermediate Representation (HIR).
///
/// The HIR is an optimized, more refined representation of the program derived from the AST.
/// While the AST closely mirrors the source syntax, the HIR represents a more semantically
/// meaningful structure that's better suited for type checking, optimization, and code generation.
///
/// Key differences between HIR and AST:
/// - Special forms (like `let`, `if`, etc.) are fully expanded in the HIR
/// - Syntactic sugar is desugared and normalized
/// - The HIR uses interned references for better memory efficiency
/// - Nodes are more compact and optimized for compiler operations
///
/// Each HIR node has:
/// - A unique identifier for tracking through compilation phases
/// - A span linking to its origin in source code for error reporting
/// - A specific kind that determines the node's semantics and structure
///
/// The HIR is designed to be immutable after construction and uses arena allocation
/// through the `'heap` lifetime for efficient memory management.
#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
pub struct Node<'heap> {
    pub id: HirId,
    pub span: SpanId,

    pub kind: NodeKind<'heap>,
}
