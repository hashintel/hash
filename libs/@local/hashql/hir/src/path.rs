use hashql_core::{intern::Interned, symbol::Ident};

mod private {
    #[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
    pub struct Marker;
}

#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
pub struct QualifiedPath<'heap>(pub Interned<'heap, [Ident<'heap>]>, pub private::Marker);

impl<'heap> QualifiedPath<'heap> {
    /// Create a new `QualifiedPath` value.
    ///
    /// The value referred to *must* follow all the constraints and guarantees that are required for
    /// the `QualifiedPath` type.
    #[inline]
    #[must_use]
    pub const fn new_unchecked(value: Interned<'heap, [Ident<'heap>]>) -> Self {
        Self(value, private::Marker)
    }
}
