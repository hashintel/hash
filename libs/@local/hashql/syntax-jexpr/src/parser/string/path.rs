use hashql_ast::node::{
    id::NodeId,
    path::{Path, PathSegment},
};
use winnow::{
    ModalResult, Parser as _,
    combinator::{delimited, opt},
    error::ParserError,
};

use super::{
    combinator::{separated_boxed1, ws},
    context::Input,
    generic::parse_generic_argument,
    ident::parse_ident,
};

fn parse_path_segment<'heap, 'span, 'source, E>(
    input: &mut Input<'heap, 'span, 'source>,
) -> ModalResult<PathSegment<'heap>, E>
where
    E: ParserError<Input<'heap, 'span, 'source>>,
{
    let context = input.state;

    let arguments = opt(delimited(
        "<",
        separated_boxed1(context.heap, parse_generic_argument, ws(",")),
        ">",
    ));

    (parse_ident, arguments)
        .with_span()
        .map(|((ident, arguments), span)| PathSegment {
            id: NodeId::PLACEHOLDER,
            span: context.span(span),
            name: ident,
            arguments: arguments.unwrap_or_else(|| context.heap.empty_slice()),
        })
        .parse_next(input)
}

pub(crate) fn parse_path<'heap, 'span, 'source, E>(
    input: &mut Input<'heap, 'span, 'source>,
) -> ModalResult<Path<'heap>, E>
where
    E: ParserError<Input<'heap, 'span, 'source>>,
{
    let context = input.state;

    let root = opt("::").map(|value| value.is_some());
    let segments = separated_boxed1(context.heap, parse_path_segment, "::");

    (root, segments)
        .with_span()
        .map(|((rooted, segments), span)| Path {
            id: NodeId::PLACEHOLDER,
            span: context.span(span),
            rooted,
            segments,
        })
        .parse_next(input)
}

#[cfg(test)]
mod tests {
    #![expect(clippy::non_ascii_literal)]
    use super::parse_path;
    use crate::parser::string::test::{bind_parser, test_cases};

    // Bind our parser to create a testing function
    bind_parser!(SyntaxDump; fn parse(parse_path));

    // TODO: tests
}
