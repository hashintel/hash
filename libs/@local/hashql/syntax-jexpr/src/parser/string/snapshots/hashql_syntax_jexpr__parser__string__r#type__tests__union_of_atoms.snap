---
source: libs/@local/hashql/syntax-jexpr/src/parser/string/type.rs
description: Union of atom types
expression: Int | String | Boolean
info:
  kind: Ok
---
Type#4294967040@10
  TypeKind (Union)
    UnionType#4294967040@10
      Type#4294967040@3
        TypeKind (Path)
          Path#4294967040@3 (rooted: false)
            PathSegment#4294967040@2 (name: Int)
      Type#4294967040@6
        TypeKind (Path)
          Path#4294967040@6 (rooted: false)
            PathSegment#4294967040@5 (name: String)
      Type#4294967040@9
        TypeKind (Path)
          Path#4294967040@9 (rooted: false)
            PathSegment#4294967040@8 (name: Boolean)
